qos/evm/endpoint.go

```diff
@@ -94,29 +97,44 @@ func (e endpoint) getBlockNumber() (uint64, error) {
 	return *e.parsedBlockNumberResponse, nil
 }
 
-func (e *endpoint) validateArchivalCheck(archivalBalance string, endpointAddr protocol.EndpointAddr) error {
-	if e.archivalBalance == "" {
+// validateArchivalCheck validates that the endpoint has returned an archival balance for the perceived block number.
+
+// If the archival check is not enabled for the service, this will always return a nil error.
+func (e *endpoint) validateArchivalCheck(archivalState archivalState) error {
+	if !archivalState.archivalCheckConfig.Enabled {
+		return nil
+	}
+
+	if e.observedArchivalBalance == "" {
 		return errNoArchivalBalanceObs
 	}
-	if e.archivalBalance != archivalBalance {
-		return fmt.Errorf(errInvalidArchivalBalanceObs, e.archivalBalance, archivalBalance)
+	if e.observedArchivalBalance != archivalState.expectedBalance {
+		return fmt.Errorf(errInvalidArchivalBalanceObs, e.observedArchivalBalance, archivalState.expectedBalance)
 	}
+
 	return nil
 }
 
+// getArchivalBalance returns the observed archival balance for the endpoint at the archival block height.
+// Returns an error if the endpoint hasn't yet returned an archival balance observation.
 func (e endpoint) getArchivalBalance() (string, error) {
-	if e.archivalBalance == "" {
+	if e.observedArchivalBalance == "" {
 		return "", errNoArchivalBalanceObs
 	}
-	return e.archivalBalance, nil
+	return e.observedArchivalBalance, nil
 }
 
 // ApplyObservation updates the data stored regarding the endpoint using the supplied observation.
 // It Returns true if the observation was not unrecognized, i.e. mutated the endpoint.
+//
+// For archival balance observations:
+// - Only updates the archival balance if the balance was observed at the specified archival block height
+// - This ensures accurate historical balance validation at the specific block number
+//
 // TODO_TECHDEBT(@adshmh): add a method to distinguish the following two scenarios:
 //   - an endpoint that returned in invalid response.
 //   - an endpoint with no/incomplete observations.
-func (e *endpoint) ApplyObservation(obs *qosobservations.EVMEndpointObservation, shouldPerformArchivalCheck bool) bool {
+func (e *endpoint) ApplyObservation(obs *qosobservations.EVMEndpointObservation, archivalBlockHeight string) bool {
```

Comment:
Ditto here: not a blocker, but a TODO that these should be moved to ServiceState eventually, to remove the need for passing around multiple data items.

Reply:
Ditto above.